usr/sbin/rear

```diff
@@ -524,27 +524,49 @@ fi
 # "exec 7>&-"
 # "exec 6<&-"
 
-# Used to determine whether TMPDIR has been changed in user config.
-# After the `mktemp` below, changing TMPDIR does not have the intended effect.
-# Save the current value to detect changes.
-saved_tmpdir="${TMPDIR-}"
+# Create a sufficiently safe temporary working area BUILD_DIR for ReaR and
+# set TMPDIR to TMP_DIR within BUILD_DIR for programs that are called by ReaR:
 if test "$WORKFLOW" != "help" ; then
     # Prepend temporary working area (BUILD_DIR) removal exit task
     # so it gets executed directly before the above listed exit tasks.
     # This must be done before the first possible call of the 'Error' function
     # otherwise we may error out but leave the build area behind:
     QuietAddExitTask cleanup_build_area_and_end_program
-    # Create temporary working area:
+    # We rely on sufficiently safe 'mktemp -d' behaviour
+    # that the BUILD_DIR directory will have read, write, and search permissions
+    # for the current user (i.e. for root), but no permissions for the group or others
+    # (i.e. BUILD_DIR is "drwx------ root root /var/tmp/rear.XXXXXXXXXXXXXXX"), cf.
+    # https://www.gnu.org/software/coreutils/manual/html_node/mktemp-invocation.html#mktemp-invocation
+    # as of this writing in Feb. 2024, see also https://github.com/rear/rear/issues/3167
     BUILD_DIR="$( mktemp -d -t rear.XXXXXXXXXXXXXXX || Error "Could not create build area '$BUILD_DIR'" )"
     ROOTFS_DIR=$BUILD_DIR/rootfs
     TMP_DIR=$BUILD_DIR/tmp
     mkdir -p $ROOTFS_DIR || Error "Could not create $ROOTFS_DIR"
     mkdir -p $TMP_DIR || Error "Could not create $TMP_DIR"
+    # For programs that are called by ReaR set TMPDIR to ReaR's TMP_DIR
+    # regardless what TMPDIR was or what it currently is, for example when
+    # TMPDIR was initially unset so it is now ReaR's default /var/tmp
+    # or when TMPDIR is exported by the user like TMPDIR=/somewhere
+    # to ensure in any case that also for programs that are called by ReaR
+    # ReaR will clean up all temporary stuff carefully and completely
+    # via the function cleanup_build_area_and_end_program()
+    # and that temporary files from programs that are called by ReaR
+    # are sufficiently safe against unwanted access by non-root users
+    # via 'mktemp -d' where BUILD_DIR and implicitly TMP_DIR=$BUILD_DIR/tmp
+    # has permissions only for root but none for the group or others,
+    # see https://github.com/rear/rear/issues/3167
+    export TMPDIR="$TMP_DIR"
```

Comment:
This will change all temporary files created by `mktemp` to be located under `$TMP_DIR`. There are not many of them, probably the most visible is the ESP mountpoint in 100_create_efiboot.sh. I don't think this is a problem, it could be even an improvement (everything is under the same place), but it is something to be aware of.